File: examples/android/CMeSdkJava/app/src/main/java/me/connect/sdk/java/sample/homepage/HomePageViewModel.java
package me.connect.sdk.java.sample.homepage;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.Executors;

import me.connect.sdk.java.Connections;
import me.connect.sdk.java.ConnectionsUtils;
import me.connect.sdk.java.Credentials;
import me.connect.sdk.java.Messages;
import me.connect.sdk.java.OutOfBandHelper;
import me.connect.sdk.java.ProofRequests;
import me.connect.sdk.java.Proofs;
import me.connect.sdk.java.StructuredMessages;
import me.connect.sdk.java.Utils;
import me.connect.sdk.java.message.Message;
import me.connect.sdk.java.message.MessageType;
import me.connect.sdk.java.message.StructuredMessageHolder;
import me.connect.sdk.java.sample.SingleLiveData;
import me.connect.sdk.java.sample.db.Database;
import me.connect.sdk.java.sample.db.entity.Action;
import me.connect.sdk.java.sample.db.entity.Connection;
import me.connect.sdk.java.sample.db.entity.CredentialOffer;
import me.connect.sdk.java.sample.db.entity.ProofRequest;
import me.connect.sdk.java.sample.db.entity.StructuredMessage;
import me.connect.sdk.java.sample.messages.CredDataHolder;
import me.connect.sdk.java.sample.messages.ProofDataHolder;

import static me.connect.sdk.java.sample.db.ActionStatus.HISTORIZED;
import static me.connect.sdk.java.sample.homepage.Results.ACTION_FAILURE;
import static me.connect.sdk.java.sample.homepage.Results.ACTION_SUCCESS;
import static me.connect.sdk.java.sample.homepage.Results.REJECT;
import static me.connect.sdk.java.sample.homepage.Results.FAILURE;
import static me.connect.sdk.java.sample.db.ActionStatus.PENDING;
import static me.connect.sdk.java.sample.homepage.Results.SUCCESS;
import static me.connect.sdk.java.sample.homepage.Results.QUESTION_SUCCESS;
import static me.connect.sdk.java.sample.homepage.Results.QUESTION_FAILURE;

public class HomePageViewModel extends AndroidViewModel {
    private final Database db;
    private LiveData<List<Action>> actions;

    public HomePageViewModel(@NonNull Application application) {
        super(application);
        db = Database.getInstance(application);
    }

    public LiveData<List<Action>> getActions() {
        if (actions == null) {
            actions = db.actionDao().getActionsByStatus(PENDING.toString());
        }
        return actions;
    }

    public SingleLiveData<Results> accept(int actionId) {
        SingleLiveData<Results> data = new SingleLiveData<>();
        acceptProcess(actionId, data);
        return data;
    }

    public SingleLiveData<Results> reject(int actionId) {
        SingleLiveData<Results> data = new SingleLiveData<>();
        rejectProcess(actionId, data);
        return data;
    }

    public SingleLiveData<Results> newAction(String invite) {
        SingleLiveData<Results> data = new SingleLiveData<>();
        createAction(invite, data);
        return data;
    }

    public SingleLiveData<Results> checkMessages() {
        SingleLiveData<Results> data = new SingleLiveData<>();
        checkAllMessages(data);
        return data;
    }

    public SingleLiveData<Results> answerMessage(int actionId, JSONObject answer) {
        SingleLiveData<Results> data = new SingleLiveData<>();
        answerStructMessage(actionId, answer, data);
        return data;
    }

    private void answerStructMessage(int actionId, JSONObject answer, SingleLiveData<Results> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            Action action = db.actionDao().getActionsById(actionId);
            StructuredMessage sm = db.structuredMessageDao().getByEntryIdAndPwDid(action.entryId, action.pwDid);

            Connection connection = db.connectionDao().getByPwDid(sm.pwDid);
            StructuredMessages.answer(connection.serialized, sm.serialized, answer).handle((res, err) -> {
                if (err == null) {
                    try {
                        sm.selectedAnswer = answer.getString("text");
                        db.structuredMessageDao().update(sm);
                        liveData.postValue(QUESTION_FAILURE);
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }
                liveData.postValue(QUESTION_SUCCESS);
                return null;
            });
            addToHistory(actionId, "Ask to question", db, liveData);
        });
    }

    private void checkAllMessages(SingleLiveData<Results> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            Messages.getAllPendingMessages().handle((messages, throwable) -> {
                if (throwable != null) {
                    throwable.printStackTrace();
                }
                liveData.postValue(SUCCESS);
                for (Message message : messages) {
                    if (MessageType.CREDENTIAL_OFFER.matches(message.getType())) {
                        credentialOffersProcess(message, liveData);
                    }
                    if (MessageType.PROOF_REQUEST.matches(message.getType())) {
                        proofRequestProcess(message, liveData);
                    }
                    if (MessageType.QUESTION.matches(message.getType())) {
                        questionsProcess(message, liveData);
                    }
                }
                return null;
            });
        });
    }

    private void credentialOffersProcess(Message message, SingleLiveData<Results> liveData) {
        CredDataHolder holder = CredDataHolder.extractDataFromCredentialsOfferMessage(message);
        String pwDid = message.getPwDid();
        Connection connection = db.connectionDao().getByPwDid(pwDid);
        try {
            Credentials.createWithOffer(UUID.randomUUID().toString(), holder.offer).handle((co, err) -> {
                if (err != null) {
                    err.printStackTrace();
                } else {
                    CredentialOffer offer = new CredentialOffer();
                    offer.threadId = holder.threadId;
                    offer.claimId = holder.id;
                    offer.pwDid = pwDid;
                    offer.serialized = co;
                    db.credentialOffersDao().insertAll(offer);

                    Messages.updateMessageStatus(pwDid, message.getUid());
                    createActionWithOffer(
                            MessageType.CREDENTIAL_OFFER.toString(),
                            holder.name,
                            connection.icon,
                            holder.id,
                            pwDid,
                            liveData
                    );
                }
                return null;
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void proofRequestProcess(Message message, SingleLiveData<Results> liveData) {
        ProofDataHolder holder = ProofDataHolder.extractRequestedFieldsFromProofMessage(message);
        String pwDid = message.getPwDid();
        Connection connection = db.connectionDao().getByPwDid(pwDid);
        try {
            Proofs.createWithRequest(UUID.randomUUID().toString(), holder.proofReq).handle((pr, err) -> {
                if (err != null) {
                    err.printStackTrace();
                } else {
                    ProofRequest proof = new ProofRequest();
                    proof.serialized = pr;
                    proof.pwDid = pwDid;
                    proof.threadId = holder.threadId;
                    db.proofRequestDao().insertAll(proof);

                    Messages.updateMessageStatus(pwDid, message.getUid());
                    createActionWithProof(
                            MessageType.CREDENTIAL_OFFER.toString(),
                            holder.name,
                            connection.icon,
                            holder.threadId,
                            liveData
                    );
                }
                return null;
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void questionsProcess(Message message, SingleLiveData<Results> liveData) {
        StructuredMessageHolder holder = StructuredMessages.extract(message);

        String pwDid = message.getPwDid();
        try {
            StructuredMessage sm = new StructuredMessage();
            sm.pwDid = pwDid;
            sm.entryId = holder.getId();
            sm.type = holder.getType();
            sm.serialized = message.getPayload();
            sm.answers = holder.getResponses();
            db.structuredMessageDao().insertAll(sm);

            if ("question".equals(holder.getType())) {
                Messages.updateMessageStatus(pwDid, message.getUid());
            }
            createActionWithQuestion(
                    MessageType.CREDENTIAL_OFFER.toString(),
                    holder.getQuestionText(),
                    holder.getQuestionDetail(),
                    pwDid,
                    holder.getId(),
                    holder.getResponses(),
                    liveData
            );
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void acceptProcess(int actionId, SingleLiveData<Results> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                Action action = db.actionDao().getActionsById(actionId);
                if (action.type == null) {
                    StateConnections.createConnection(action, db, liveData);
                    return;
                }
                if (action.type.equals(MessageType.CREDENTIAL_OFFER.toString())) {
                    CredentialOffer offer = db.credentialOffersDao().getByPwDidAndClaimId(action.claimId, action.pwDid);
                    StateCredentialOffers.acceptCredentialOffer(offer, db, liveData, action);
                    return;
                }
                if (action.type.equals(MessageType.PROOF_REQUEST.toString())) {
                    ProofRequest proof = db.proofRequestDao().getByThreadId(action.threadId);
                    StateProofRequests.acceptProofReq(proof, db, liveData, action);
                }
            } catch (Exception e) {
                e.printStackTrace();
                liveData.postValue(FAILURE);
            }
        });
    }

    private void rejectProcess(int actionId, SingleLiveData<Results> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                Action action = db.actionDao().getActionsById(actionId);
                if (action.type == null) {
                    addToHistory(actionId, "Rejected", db, liveData);
                    return;
                }
                if (action.type.equals(MessageType.CREDENTIAL_OFFER.toString())) {
                    CredentialOffer offer = db.credentialOffersDao().getByPwDidAndClaimId(action.claimId, action.pwDid);
                    StateCredentialOffers.rejectCredentialOffer(offer, db, liveData);
                    return;
                }
                if (action.type.equals(MessageType.PROOF_REQUEST.toString())) {
                    ProofRequest proof = db.proofRequestDao().getByThreadId(action.threadId);
                    StateProofRequests.rejectProofReq(proof, db, liveData);
                }
            } catch (Exception e) {
                e.printStackTrace();
                liveData.postValue(FAILURE);
            }
        });
    }

    private void createAction(String invite, SingleLiveData<Results> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                String parsedInvite = ConnectionsUtils.parseInvite(invite);
                Connections.InvitationType invitationType = Connections.getInvitationType(parsedInvite);
                JSONObject inviteObject = new JSONObject(parsedInvite);
                JSONArray attach = inviteObject.getJSONArray("request~attach");

                if (ConnectionsUtils.isOutOfBandType(invitationType) && attach.length() == 0) {
                    Action action = new Action();
                    action.invite = invite;
                    action.name = inviteObject.getString("label");
                    action.description = inviteObject.getString("goal");
                    action.icon = inviteObject.getString("profileUrl");
                    action.status = HISTORIZED.toString();
                    db.actionDao().insertAll(action);
                    StateConnections.createConnection(action, db, liveData);
                } else if (ConnectionsUtils.isOutOfBandType(invitationType) && attach.length() != 0) {
                    String extractedAttachRequest = OutOfBandHelper.extractRequestAttach(parsedInvite);
                    JSONObject attachRequestObject = Utils.convertToJSONObject(extractedAttachRequest);

                    assert attachRequestObject != null;
                    String attachType = attachRequestObject.getString("@type");

                    if (ConnectionsUtils.isCredentialInviteType(attachType)) {
                        JSONObject preview = attachRequestObject.getJSONObject("credential_preview");
                        Action action = new Action();
                        action.invite = invite;
                        action.name = attachRequestObject.getString("comment");
                        action.description = inviteObject.getString("goal");
                        action.details = preview.getJSONArray("attributes").getString(0);
                        action.icon = inviteObject.getString("profileUrl");
                        action.status = PENDING.toString();
                        db.actionDao().insertAll(action);
                    }
                    if (ConnectionsUtils.isProofInviteType(attachType)) {
                        JSONObject decodedProofAttach = ProofRequests.decodeProofRequestAttach(attachRequestObject);

                        Action action = new Action();
                        action.invite = invite;
                        action.name = ProofRequests.extractRequestedNameFromProofRequest(decodedProofAttach);
                        action.description = inviteObject.getString("goal");
                        action.details = ProofRequests.extractRequestedAttributesFromProofRequest(decodedProofAttach);
                        action.icon = inviteObject.getString("profileUrl");
                        action.status = PENDING.toString();
                        db.actionDao().insertAll(action);
                    }
                } else {
                    Action action = new Action();
                    action.invite = invite;
                    action.name = inviteObject.getString("label");
                    action.description = inviteObject.getString("goal");
                    action.icon = inviteObject.getString("profileUrl");
                    action.status = PENDING.toString();
                    db.actionDao().insertAll(action);
                }
                liveData.postValue(ACTION_SUCCESS);
            } catch (JSONException e) {
                e.printStackTrace();
                liveData.postValue(ACTION_FAILURE);
            }
        });
    }

    private void createActionWithOffer(
            String type,
            String name,
            String icon,
            String offerId,
            String pwDid,
            SingleLiveData<Results> liveData
    ) {
        try {
            Action action = new Action();
            action.type = type;
            action.name = name;
            action.description = "To issue the credential";
            action.icon = icon;
            action.claimId = offerId;
            action.pwDid = pwDid;
            action.status = PENDING.toString();
            db.actionDao().insertAll(action);

            liveData.postValue(SUCCESS);
        } catch (Exception e) {
            e.printStackTrace();
            liveData.postValue(FAILURE);
        }
    }

    private void createActionWithProof(
            String type,
            String name,
            String icon,
            String threadId,
            SingleLiveData<Results> liveData
    ) {
        try {
            Action action = new Action();
            action.type = type;
            action.name = name;
            action.description = "Share the proof";
            action.icon = icon;
            action.threadId = threadId;
            action.status = PENDING.toString();
            db.actionDao().insertAll(action);

            liveData.postValue(SUCCESS);
        } catch (Exception e) {
            e.printStackTrace();
            liveData.postValue(FAILURE);
        }
    }

    private void createActionWithQuestion(
            String type,
            String name,
            String details,
            String pwDid,
            String entryId,
            List<StructuredMessageHolder.Response> messageAnswers,
            SingleLiveData<Results> liveData
    ) {
        try {
            Action action = new Action();
            action.invite = null;
            action.type = type;
            action.name = name;
            action.description = "Answer the questions";
            action.details = details;
            action.pwDid = pwDid;
            action.entryId = entryId;
            action.messageAnswers = messageAnswers;
            action.status = PENDING.toString();
            db.actionDao().insertAll(action);

            liveData.postValue(SUCCESS);
        } catch (Exception e) {
            e.printStackTrace();
            liveData.postValue(FAILURE);
        }
    }

    public static void addHistoryAction(
            Database db,
            String name,
            String description,
            String icon,
            SingleLiveData<Results> liveData
    ) {
        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                Action action = new Action();
                action.invite = null;
                action.name = name;
                action.description = description;
                action.icon = icon;
                action.status = HISTORIZED.toString();
                db.actionDao().insertAll(action);

                liveData.postValue(ACTION_SUCCESS);
            } catch (Exception e) {
                e.printStackTrace();
                liveData.postValue(ACTION_FAILURE);
            }
        });
    }

    public static void addToHistory(int actionId, String description, Database db, SingleLiveData<Results> liveData) {
        try {
            Action action = db.actionDao().getActionsById(actionId);
            action.status = HISTORIZED.toString();
            action.description = description;
            db.actionDao().update(action);
            liveData.postValue(REJECT);
        } catch (Exception e) {
            e.printStackTrace();
            liveData.postValue(FAILURE);
        }
    }
}

File: examples/android/CMeSdkJava/app_kt/src/main/kotlin/me/connect/sdk/java/samplekt/homepage/HomePageViewModel.kt
package me.connect.sdk.java.samplekt.homepage

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.future.await
import kotlinx.coroutines.launch
import me.connect.sdk.java.*
import me.connect.sdk.java.message.Message
import me.connect.sdk.java.message.MessageType
import me.connect.sdk.java.message.StructuredMessageHolder
import me.connect.sdk.java.samplekt.SingleLiveData
import me.connect.sdk.java.samplekt.db.ActionStatus.*
import me.connect.sdk.java.samplekt.db.Database
import me.connect.sdk.java.samplekt.db.entity.*
import me.connect.sdk.java.samplekt.homepage.Results.*
import me.connect.sdk.java.samplekt.messages.CredDataHolder
import me.connect.sdk.java.samplekt.messages.ProofDataHolder
import me.connect.sdk.java.samplekt.wrap
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class HomePageViewModel(application: Application) : AndroidViewModel(application) {
    private val db: Database = Database.getInstance(application)
    private val actionsLiveData by lazy {
        db.actionDao().getActionsByStatus(PENDING.toString())
    }

    fun getActions(): LiveData<List<Action>> = actionsLiveData

    fun accept(actionId: Int): SingleLiveData<Results> {
        val data = SingleLiveData<Results>()
        acceptProcess(actionId, data)
        return data
    }

    fun reject(actionId: Int): SingleLiveData<Results> {
        val data = SingleLiveData<Results>()
        rejectProcess(actionId, data)
        return data
    }

    fun newAction(invite: String): SingleLiveData<Results> {
        val data = SingleLiveData<Results>()
        createAction(invite, data)
        return data
    }

    fun checkMessages(): SingleLiveData<Results> {
        val data = SingleLiveData<Results>()
        checkAllMessages(data)
        return data
    }

    fun answerMessage(actionId: Int, answer: String): SingleLiveData<Results> {
        val data = SingleLiveData<Results>()
        answerStructMessage(actionId, answer, data)
        return data
    }

    private fun answerStructMessage(actionId: Int, answer: String, liveData: SingleLiveData<Results>) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val action = db.actionDao().getActionsById(actionId)
            val con = db.connectionDao().getByPwDid(action.pwDid!!)
            val sm = db.structuredMessageDao().getByEntryIdAndPwDid(action.entryId, action.pwDid)

            StructuredMessages.answer(con.serialized, sm!!.messageId, sm.type, sm.serialized, answer).wrap().await()
            sm.selectedAnswer = answer;
            db.structuredMessageDao().update(sm)
            liveData.postValue(QUESTION_SUCCESS)
            HistoryActions.addToHistory(actionId, "Ask to question", db, liveData);
        } catch (e: Exception) {
            e.printStackTrace()
            liveData.postValue(QUESTION_FAILURE);
        }
    }

    private fun checkAllMessages(liveData: SingleLiveData<Results>) = viewModelScope.launch(Dispatchers.IO) {
        val messages = Messages.getAllPendingMessages().wrap().await()
        liveData.postValue(SUCCESS)
        for (message in messages) {
            if (MessageType.CREDENTIAL_OFFER.matches(message.type)) {
                credentialOffersProcess(message, liveData)
            }
            if (MessageType.PROOF_REQUEST.matches(message.type)) {
                proofRequestProcess(message, liveData)
            }
            if (MessageType.QUESTION.matches(message.type)) {
                questionsProcess(message, liveData)
            }
        }
    }

    private suspend fun credentialOffersProcess(message: Message, liveData: SingleLiveData<Results>) {
        try {
            val holder = CredDataHolder.extractDataFromCredentialsOfferMessage(message)
            val pwDid: String = message.pwDid
            val connection: Connection = db.connectionDao().getByPwDid(pwDid)
            val co = Credentials.createWithOffer(UUID.randomUUID().toString(), holder!!.offer).wrap().await()
            val offer = CredentialOffer(
                claimId = holder.id,
                name = holder.name,
                pwDid = pwDid,
                attributes = holder.attributes,
                serialized = co,
                messageId = message.uid,
                attachConnectionLogo = connection.icon
            )
            db.credentialOffersDao().insertAll(offer)

            Messages.updateMessageStatus(pwDid, message.uid)

            createActionWithOffer(
                MessageType.CREDENTIAL_OFFER.toString(),
                holder.name,
                connection.icon!!,
                holder.id,
                pwDid,
                liveData
            );
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private suspend fun proofRequestProcess(message: Message, liveData: SingleLiveData<Results>) {
        try {
            val holder = ProofDataHolder.extractRequestedFieldsFromProofMessage(message)!!
            val pwDid: String = message.pwDid
            val connection: Connection = db.connectionDao().getByPwDid(pwDid)
            val pr = Proofs.createWithRequest(UUID.randomUUID().toString(), holder.proofReq).wrap().await()
            val proof = ProofRequest(
                serialized = pr,
                name = holder.name,
                pwDid = pwDid,
                attributes = holder.attributes,
                threadId = holder.threadId,
                attachConnectionLogo = connection.icon,
                messageId = message.uid
            )
            db.proofRequestDao().insertAll(proof)

            Messages.updateMessageStatus(pwDid, message.uid)

            createActionWithProof(
                MessageType.CREDENTIAL_OFFER.toString(),
                holder.name,
                connection.icon!!,
                holder.threadId,
                liveData
            );
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private suspend fun questionsProcess(
        message: Message,
        liveData: SingleLiveData<Results>
    ) {
        val holder = StructuredMessages.extract(message)
        val pwDid = message.pwDid
        val connection = db.connectionDao().getByPwDid(pwDid)
        try {
            val sm = StructuredMessage(
                pwDid = pwDid,
                messageId = message.uid,
                entryId = holder.id,
                questionText = holder.questionText,
                questionDetail = holder.questionDetail,
                type = holder.type,
                serialized = message.payload,
                answers = holder.responses
            )

            db.structuredMessageDao().insertAll(sm)
            if ("question" == holder.type) {
                Messages.updateMessageStatus(pwDid, message.uid)
            }
            createActionWithQuestion(
                MessageType.CREDENTIAL_OFFER.toString(),
                connection.name,
                connection.icon!!,
                pwDid,
                holder.id,
                holder.responses,
                liveData
            )
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
    }

    private fun acceptProcess(
        actionId: Int,
        liveData: SingleLiveData<Results>
    ) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val action = db.actionDao().getActionsById(actionId)
            if (action.type == null) {
                StateConnections.createConnection(action, db, liveData)
                return@launch
            }
            if (action.type == MessageType.CREDENTIAL_OFFER.toString()) {
                val offer = db.credentialOffersDao().getByPwDidAndClaimId(action.claimId, action.pwDid)
                StateCredentialOffers.acceptCredentialOffer(offer!!, db, liveData, action)
                return@launch
            }
            if (action.type == MessageType.PROOF_REQUEST.toString()) {
                val proof = db.proofRequestDao().getByThreadId(action.threadId)
                StateProofRequests.acceptProofReq(proof!!, db, liveData, action)
            }
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
            liveData.postValue(FAILURE)
        }
    }

    private fun rejectProcess(
        actionId: Int,
        liveData: SingleLiveData<Results>
    ) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val action = db.actionDao().getActionsById(actionId)
            if (action.type == null) {
                HistoryActions.addToHistory(actionId, "Rejected", db, liveData)
                return@launch
            }
            if (action.type == MessageType.CREDENTIAL_OFFER.toString()) {
                val offer = db.credentialOffersDao().getByPwDidAndClaimId(action.claimId, action.pwDid)
                StateCredentialOffers.rejectCredentialOffer(offer!!, db, liveData)
                return@launch
            }
            if (action.type == MessageType.PROOF_REQUEST.toString()) {
                val proof = db.proofRequestDao().getByThreadId(action.threadId)
                StateProofRequests.rejectProofReq(proof!!, db, liveData)
            }
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
            liveData.postValue(FAILURE)
        }

    }

    private fun createAction(
        invite: String,
        liveData: SingleLiveData<Results>
    ) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val parsedInvite = ConnectionsUtils.parseInvite(invite)
            val invitationType =
                Connections.getInvitationType(parsedInvite)
            val inviteObject = JSONObject(parsedInvite)
            val attach = inviteObject.getJSONArray("request~attach")

            if (ConnectionsUtils.isOutOfBandType(invitationType) && attach.length() == 0) {
                val action = Action(
                    invite = invite,
                    name = inviteObject.getString("label"),
                    description = inviteObject.getString("goal"),
                    icon = inviteObject.getString("profileUrl"),
                    status = HISTORIZED.toString()
                )

                db.actionDao().insertAll(action)
                StateConnections.createConnection(action, db, liveData)
            } else if (ConnectionsUtils.isOutOfBandType(invitationType) && attach.length() != 0) {
                val extractedAttachRequest = OutOfBandHelper.extractRequestAttach(parsedInvite)
                val attachRequestObject = Utils.convertToJSONObject(extractedAttachRequest)!!
                val attachType = attachRequestObject.getString("@type")
                if (ConnectionsUtils.isCredentialInviteType(attachType)) {
                    val preview = attachRequestObject.getJSONObject("credential_preview")
                    val action = Action(
                        invite = invite,
                        name = attachRequestObject.getString("comment"),
                        description = inviteObject.getString("goal"),
                        details = preview.getJSONArray("attributes").getString(0)!!,
                        icon = inviteObject.getString("profileUrl"),
                        status = PENDING.toString()
                    )
                    db.actionDao().insertAll(action)
                    liveData.postValue(ACTION_SUCCESS)
                    return@launch
                }
                if (ConnectionsUtils.isProofInviteType(attachType)) {
                    val decodedProofAttach = ProofRequests.decodeProofRequestAttach(attachRequestObject)
                    val action = Action(
                        invite = invite,
                        name = ProofRequests.extractRequestedNameFromProofRequest(decodedProofAttach),
                        description = inviteObject.getString("goal"),
                        details = ProofRequests.extractRequestedAttributesFromProofRequest(decodedProofAttach),
                        icon = inviteObject.getString("profileUrl"),
                        status = PENDING.toString()
                    )

                    db.actionDao().insertAll(action)
                    liveData.postValue(ACTION_SUCCESS)
                    return@launch
                }
            } else {
                val action = Action(
                    invite = invite,
                    name = inviteObject.getString("label"),
                    description = inviteObject.getString("goal"),
                    icon = inviteObject.getString("profileUrl"),
                    status = PENDING.toString()
                )

                db.actionDao().insertAll(action)
                liveData.postValue(ACTION_SUCCESS)
            }
        } catch (e: JSONException) {
            e.printStackTrace()
            liveData.postValue(ACTION_FAILURE)
        }
    }

    private suspend fun createActionWithOffer(
        type: String,
        name: String,
        icon: String,
        offerId: String,
        pwDid: String,
        liveData: SingleLiveData<Results>
    ) {
        try {
            val action = Action(
                type = type,
                name = name,
                description = "To issue the credential",
                icon = icon,
                claimId = offerId,
                pwDid = pwDid,
                status = PENDING.toString()
            )

            db.actionDao().insertAll(action)
            liveData.postValue(SUCCESS)
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
            liveData.postValue(FAILURE)
        }
    }

    private suspend fun createActionWithProof(
        type: String,
        name: String,
        icon: String,
        threadId: String,
        liveData: SingleLiveData<Results>
    ) {
        try {
            val action = Action(
                type = type,
                name = name,
                description = "Share the proof",
                icon = icon,
                threadId = threadId,
                status = PENDING.toString()
            )

            db.actionDao().insertAll(action)
            liveData.postValue(SUCCESS)
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
            liveData.postValue(FAILURE)
        }
    }

    private suspend fun createActionWithQuestion(
        type: String,
        name: String,
        icon: String,
        pwDid: String,
        entryId: String,
        messageAnswers: List<StructuredMessageHolder.Response>,
        liveData: SingleLiveData<Results>
    ) {
        try {
            val action = Action(
                invite = null,
                type = type,
                name = name,
                description = "Answer the questions",
                icon = icon,
                pwDid = pwDid,
                entryId = entryId,
                messageAnswers = messageAnswers,
                status = PENDING.toString()
            )

            db.actionDao().insertAll(action)
            liveData.postValue(SUCCESS)
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
            liveData.postValue(FAILURE)
        }
    }

    object HistoryActions {
        suspend fun addHistoryAction(
            db: Database,
            name: String,
            description: String,
            icon: String,
            liveData: SingleLiveData<Results>
        ) {
            try {
                val action = Action(
                    invite = null,
                    name = name,
                    description = description,
                    icon = icon,
                    status = HISTORIZED.toString()
                )

                db.actionDao().insertAll(action)
                liveData.postValue(ACTION_SUCCESS)
            } catch (e: java.lang.Exception) {
                e.printStackTrace()
                liveData.postValue(ACTION_FAILURE)
            }
        }

        suspend fun addToHistory(
            actionId: Int,
            description: String,
            db: Database,
            liveData: SingleLiveData<Results>
        ) {
            try {
                val action = db.actionDao().getActionsById(actionId)
                action.status = HISTORIZED.toString()
                action.description = description
                db.actionDao().update(action)
                liveData.postValue(REJECT)
            } catch (e: java.lang.Exception) {
                e.printStackTrace()
                liveData.postValue(FAILURE)
            }
        }
    }
}
